src/AWS.Deploy.CLI/ServerMode/Services/SessionAWSResourceQuery.cs

```diff
@@ -0,0 +1,378 @@
+    /// a new instance of this class will be created with a new cache.
+    /// </summary>
+    public class SessionAWSResourceQuery : IAWSResourceQueryer
+    {
+        private readonly IAWSResourceQueryer _awsResourceQueryer;
+        private readonly ConcurrentDictionary<string, object?> _cachedResponse = new();
+
+        /// <summary>
+        /// Construct an instance of SessionAWSResourceQuery. This will create an instance of AWSResourceQueryer
+        /// using the services in the IServiceProvider.
+        /// </summary>
+        /// <param name="services"></param>
+
+        public SessionAWSResourceQuery(IServiceProvider services)
+        {
+            _awsResourceQueryer = ActivatorUtilities.CreateInstance<AWSResourceQueryer>(services);
+        }
+
+        /// <summary>
+        /// Construct an instance of SessionAWSResourceQuery wrapping the passed in IAWSResourceQueryer.
+        /// This constructor is primary included for unit testing and injecting a Mock version of IAWSResourceQueryer.
+        ///
+        /// The factory Create method with a private constructor is used to avoid this constructor being attempted
+        /// to be used in DI resolution.
+        /// </summary>
+        /// <param name="awsResourceQueryer"></param>
+        private SessionAWSResourceQuery(IAWSResourceQueryer awsResourceQueryer)
+        {
+            _awsResourceQueryer = awsResourceQueryer;
+        }
+
+        /// <summary>
+        /// Construct an instance of SessionAWSResourceQuery wrapping the passed in IAWSResourceQueryer.
+        /// This method is primary included for unit testing and injecting a Mock version of IAWSResourceQueryer.
+        /// </summary>
+        /// <param name="awsResourceQueryer"></param>
+        public static IAWSResourceQueryer Create(IAWSResourceQueryer awsResourceQueryer)
+        {
+            return new SessionAWSResourceQuery(awsResourceQueryer);
+        }
+
+        private string CreateCacheKey(IEnumerable<object?>? args = null, [CallerMemberName] string caller = "")
+        {
+            var values = new List<object>
+            {
+                caller
+            };
+            if (args != null)
+            {
+                foreach (var item in args)
+                {
+                    if (item is null)
+                    {
+                        values.Add("null");
+                    }
+                    if (item is System.Collections.IEnumerable itemEnumerable)
+                    {
+                        values.Add("IEnumerable");
+                        foreach (var subItem in itemEnumerable)
+                        {
+                            if (subItem is null)
+                                values.Add("null");
+                            else
+                                values.Add(subItem.ToString() ?? "null");
+                        }
+                    }
```

Comment:
Checking my understanding - for something like `GetS3BucketLocation("my-bucket")`, should this be `"GetS3BucketLocation,mybucket"`? I'm seeing `"GetS3BucketLocation,IEnumerable,m,y,-,b,u,c,k,e,t"`.

Should there be an `item.ToString()` for the case where `item` is not `null` && not an `IEnumerable`?

Reply:
Also, wonder if we should be throwing an exception for if we fail to add something to `values` for a given `item`. 

I think it'd be rare for us to add new type of signature to the interface, but if we do I think it'd be tough to remember to make sure it works here.